test/integration/sessions/sessions.test.js

```diff
@@ -72,84 +72,59 @@ describe('Sessions Spec', function () {
+            return client.db('test').collection('foo').find({}, { session }).toArray();
           }
-        ].forEach(testCase => {
-          it(testCase.description, function () {
-            const client = test.client;
-
-            return client
-              .withSession(testCase.operation(client))
-              .then(
-                () => expect(client.topology.s.sessionPool.sessions).to.have.length(1),
-                () => expect(client.topology.s.sessionPool.sessions).to.have.length(1)
-              )
-              .then(() => client.close())
-              .then(() => {
-                // verify that the `endSessions` command was sent
-                const lastCommand = test.commands.started[test.commands.started.length - 1];
-                expect(lastCommand.commandName).to.equal('endSessions');
-                expect(client.topology).to.not.exist;
-              });
-          });
-        });
+        },
+        {
+          description: 'should support async operations',
+          operation: client => session => async () =>
+            await client.db('test').collection('foo').find({}, { session }).toArray()
+        },
+        {
+          description: 'should support errors thrown from async operations',
+          operation: () => () => async () => {
+            throw new Error('thrown from async function');
+          }
+        },
+        {
+          description: 'should support operations that return rejected promises',
+          operation: (/* client */) => (/* session */) => {
+            return Promise.reject(new Error('something awful'));
+          }
+        },
+        {
+          description: "should support operations that don't return promises",
+          operation: (/* client */) => (/* session */) => {
+            setTimeout(() => {});
+          }
+        },
+        {
+          description: 'should support operations that throw exceptions',
+          operation: (/* client */) => (/* session */) => {
+            throw new Error('something went wrong!');
+          }
+        }
+      ];
 
-        it('supports passing options to ClientSession', function () {
+      for (const testCase of tests) {
+        it(testCase.description, function () {
           const client = test.client;
 
-          const promise = client.withSession({ causalConsistency: false }, session => {
-            expect(session.supports.causalConsistency).to.be.false;
-            return client.db('test').collection('foo').find({}, { session }).toArray();
-          });
-
-          return promise
-            .then(() => expect(client.topology.s.sessionPool.sessions).to.have.length(1))
+          return client
+            .withSession(testCase.operation(client))
+            .then(
+              () => expect(client.topology.s.sessionPool.sessions).to.have.length(1),
+              () => expect(client.topology.s.sessionPool.sessions).to.have.length(1)
+            )
             .then(() => client.close())
```

Comment:
the client close here and in the test case below should probably be in an afterEach; the test cases should also differentiate `resolve` from `reject` cases in their description and corresponding assertions; the test case descriptions should indicate that we care that sessions are cleaned up in both cases